cmd/storage-rest-server.go

```diff
@@ -731,11 +731,9 @@ type closeNotifier struct {
 
 func (c *closeNotifier) Read(p []byte) (n int, err error) {
 	n, err = c.rc.Read(p)
-	if err != nil {
-		if c.done != nil {
-			xioutil.SafeClose(c.done)
-			c.done = nil
-		}
+	if c.done != nil {
+		xioutil.SafeClose(c.done)
+		c.done = nil
```

Comment:
You are closing after the first read returns. The promise is

```
// closeNotifier is itself a ReadCloser that will notify when either an error occurs or
// the Close() function is called.
```

This will break http requests if they start writing before the request body has been read.

Reply:
c.rc.Read() is already called basically the first attempt to Read() the body.

Reply:
I will add a test to show it works and why it doesn't without this change.